src/main/java/org/marketplace/services/AdvertisementImageManagementService.java

```diff
@@ -46,4 +59,24 @@ public void deleteAdvertisementImage(Long id) {
     public List<AdvertisementImage> getAllAdvertisementImages(Long advertisementId) {
         return advertisementImageManagementRepository.findByAdvertisement_Id(advertisementId);
     }
+
+    @Transactional
+    public AdvertisementImage save(AdvertisementImage document) {
+        AdvertisementImage savedDocument = advertisementImageManagementRepository.save(document);
+        MutableAcl acl = setUpAcl(savedDocument);
+        String username = SecurityContextHolder.getContext().getAuthentication().getName();
+        acl.insertAce(0, BasePermission.ADMINISTRATION, new PrincipalSid(username), true);
```

Comment:
gotcha, initially i took it in the way that we want to enable image owner to view/modify it and ban the rest.
if i understand it correctly, if we want to enable all users for read it should be just enough to add like 
```
@PreAuthorize("hasAuthority('READ')")
@GetMapping("/{id}")
public Response<Category> requestGetCategoryById
```

Reply:
Yeah I have a feeling that the ACL might not be needed here, but your suggestion is more valuable than how I feel about it.

The main idea behind what I wanted when implementing the authorization, was to allow ALL users to hit `GET` requests, as they do not modify the main resources (Advertisement/Category/Users/Adv.Img).

However, for certain cases, we want to forbid user from hitting the `PUT/DELETE` endpoints - if they are not the owner of the resource they want to modify or they are not the Admin of our application. 

Using your snippet as an example:
```
@PreAuthorize("hasAuthority('READ')")
@GetMapping("/{id}")
public Response<Category> requestGetCategoryById
```
We would want to allow the user to hit the get category endpoint as a default, because thats what the frontend side of the app would do - we need to somehow fetch the data to display on the screen, and the same goes for Advertisement or the User. Getting the data is okay for everyone. Modifying on the other hand, not really. 

So the issue lies within adding the authorities using ACL - we would have to add the authorities for all newly created entities that are related between each other. In our application, this would be all of our entities. 

So when creating a new user, we would have to map over all Advertisements, all Images, and all Categories and give him a READ access. 

When creating a new advertisement, we need to give read access to everybody, and also give DELETE/WRITE access to user-owner and admin. 

(Same goes for AdvertisementID, and Category)

And I feel like this complicates the usage a bit. However, I cannot evaluate what is more calculation-extensive. But checking if `user.id == resource.user.id || user.role == "ADMIN"` seems more intuitive to me.

What do you think?

Reply:
it sounds like each resource requires some extra mapping like @OneToMany with User object, we can do something like that for our project, that's okay i think 👍
in real life project i would refrain from adding such dependancy, if myself being a user "some_user" and i pull out image that has other 100 users in the db who can modify it, we don't really want "some_user" to have an info about these other 100 users who can modify an image. So in other words we give additional metadata to the "image" object that we use as a hack for authentication management/ACL. 
So we kind of mess up here principal object/cookies (that stands for current session) and image which is a table in db.  
We can also think through this issue from the perspective of future extension, every new object by default will join User table, so potential change in User object will hit our business logic.
So when it comes to object modification there is no other way around then to use ACL.
Just for the sake of time we can not use ACL here, but in real life situation i would suggest to use it.

Reply:
Ohhh wow, this is actually such a great point. Hadnt given a thought about this case. 
To be honest? I am stunned that I couldnt see through that this is some great vulnerability. So yeah, lets say it was not image but something more sensitive and we are donezo basically exposing ourselves to abuse.
I will see if there is time to go forward with acl then (I might be ready with other features and might give it a try).

Great point and I will keep it in mind for future projects. Yeah - wow.

Reply:
> Ohhh wow, this is actually such a great point. Hadnt given a thought about this case. To be honest? I am stunned that I couldnt see through that this is some great vulnerability. So yeah, lets say it was not image but something more sensitive and we are donezo basically exposing ourselves to abuse. I will see if there is time to go forward with acl then (I might be ready with other features and might give it a try).
> 
> Great point and I will keep it in mind for future projects. Yeah - wow.

that's no worries, no rush with ACL, it's fine if we don't implement it, so that modification will be based on roles only